File: 129_Single Number II.py
"""
129. Single Number II
Every element appears three times except for one. Use bit counting per bit and mod 3.
Handles negative numbers by reconstructing two's complement.
"""
from typing import List

def single_number_ii(nums: List[int]) -> int:
    res = 0
    for i in range(32):
        s = 0
        for n in nums:
            s += (n >> i) & 1
        if s % 3:
            res |= (1 << i)
    # convert to signed
    if res >= 2**31:
        res -= 2**32
    return res

if __name__ == '__main__':
    print('singleNumberII([2,2,3,2]) =', single_number_ii([2,2,3,2]))  # 3
    print('singleNumberII([0,1,0,1,0,1,99]) =', single_number_ii([0,1,0,1,0,1,99]))  # 99

# Time Complexity: O(32*n) = O(n)
# Space Complexity: O(1)

File: 129_Single Number II.cpp
/*
129. Single Number II
Given an array of integers, every element appears three times except for one which appears once. Find that single one.
Use bitwise counting: sum bits modulo 3.
*/
#include <bits/stdc++.h>
using namespace std;

int singleNumberII(vector<int>& nums) {
    int result = 0;
    for (int i = 0; i < 32; ++i) {
        int sum = 0;
        for (int n : nums) sum += ( (n >> i) & 1 );
        if (sum % 3) result |= (1 << i);
    }
    return result;
}

int main() {
    vector<int> a{2,2,3,2};
    cout << "singleNumberII([2,2,3,2]) = " << singleNumberII(a) << "\n"; // 3
    vector<int> b{0,1,0,1,0,1,99};
    cout << "singleNumberII(...,99) = " << singleNumberII(b) << "\n"; // 99
    return 0;
}

// Time Complexity: O(32*n) = O(n)
// Space Complexity: O(1)
